### PR TITLE
fby35: cl: IPMI command to access DIMM device memory

### PR DESCRIPTION
**Pull request by DelphineChiu**
Summary:
- Support BMC to read SPD information and write/read PMIC register.
- IPMI command formate request:
Byte 0:2 - Byte 1:3 – MFG ID – 00A015h, LS byte first
Byte 3 - DIMM location
  00h - A0
  01h - A2
  02h - A3
  03h - A4
  04h - A6
  05h - A7
Byte 4 - Device type
  00h - SPD
  01h - SPD NVM
  02h - PMIC
Byte 5 - Read data length
if Byte 4 = 0x01 SPD NVM
  Byte 6:7 - 2 bytes offset, LS byte first
  Byte 8:~ -  Write data
Otherwise
  Byte 6 - 1 bytes offset
  Byte 7:~ - Write data

Dependency: #801 

Test plan:
- Build code: Pass
- Read SPD register: Pass
- Read SPD NVM: Pass
- Read PMIC register: Pass
- Write PMIC register: Pass

Log:
1. Using IPMI command to read SPD register.
- DIMM A0 Device Type
root@bmc-oob:~# bic-util slot3 0xE0 0xB1 0x15 0xA0 0x00 0x01 0x00 2 0x00
15 A0 00 51 18

2. Using IPMI command to read SPD NVM.
root@bmc-oob:~# dimm-util slot3 --config --dimm A2
FRU: slot3
DIMM A2:
        Type: DDR5 SDRAM
        Size: 32768 MB
        Speed: 4800 MT/s
        Manufacturer: Samsung
        Manufacturing Date: 2022 Week36
        Part Number: M321R4GA3BB6-CQKEG
        Serial Number: 039B1E95
        Register Vendor: IDT
        PMIC Vendor: MPS

- DIMM A2 Serial Number
root@bmc-oob:~# bic-util slot3 0xE0 0xB1 0x15 0xA0 0x00 0x01 0x01 4 0x05 0x02
15 A0 00 03 9B 1E 95

3. Using IPMI command to read PMIC register.
- DIMM A2 Vender ID
root@bmc-oob:~# bic-util slot3 0xE0 0xB1 0x15 0xA0 0x00 0x01 0x02 2 0x31 0x00
15 A0 00 48 E0

4. Using IPMI command to write PMIC register.
- DIMM A2 SWAout_OV (critical error)
root@bmc-oob:~# bic-util slot3 0xE0 0xB1 0x15 0xA0 0x00 0x01 0x02 2 0x35 0x90
15 A0 00 90 00
root@bmc-oob:~# log-util --print slot3
2023 Jan 09 04:37:57 log-util: User cleared FRU: 3 logs
3    slot3    2023-01-09 04:49:04    ipmid            SEL Entry: FRU: 3, Record: Standard (0x02), Time: 2023-01-09 04:49:04, Sensor: PMIC_ERR (0xB4), Event Data: (010000) DIMM A2 SWAout OV Assertion

- DIMM A2 high_temp (warning error)
root@bmc-oob:~# bic-util slot3 0xE0 0xB1 0x15 0xA0 0x00 0x01 0x02 2 0x35 0x83
15 A0 00 83 00
root@bmc-oob:~# log-util --print slot3
2023 Jan 09 04:37:57 log-util: User cleared FRU: 3 logs
3    slot3    2023-01-09 04:48:28    ipmid            SEL Entry: FRU: 3, Record: Standard (0x02), Time: 2023-01-09 04:48:28, Sensor: PMIC_ERR (0xB4), Event Data: (010C00) DIMM A2 High temp warning Assertion

### CODE REVIEW

**Comment by DelphineChiu** [comment]
Dependency: #801

File: common/dev/i3c_dimm.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		sval->integer = (data[0] * PMIC_TOTAL_POWER_MW / 1000) & 0xFFFF;
+		sval->fraction = (data[0] * PMIC_TOTAL_POWER_MW % 1000) & 0xFFFF;
+		break;
+	case DIMM_SPD_TEMP:
+		ret = pal_get_spd_temp(sensor_num, (uint8_t *)&data);
+		if (ret != 0) {
+			return SENSOR_FAIL_TO_ACCESS;
+		}
+
+		/* Accround SPEC 3.2.2 Thermal Sensor Registers Read Out Mechanism (p.105)
```
Comment (garnermic) [inline common/dev/i3c_dimm.c:65]: "Accround"?  Any idea what was meant by this.

File: meta-facebook/yv35-cl/src/platform/plat_dimm.c [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+}
+
+int pal_get_spd_temp(uint8_t sensor_num, uint8_t *data)
+{
+	CHECK_NULL_ARG_WITH_RETURN(data, -1);
+
+	int dimm_id = -1;
+
+	dimm_id = sensor_num_map_dimm_id(sensor_num);
+	if (dimm_id < 0) {
```
Comment (garnermic) [inline meta-facebook/yv35-cl/src/platform/plat_dimm.c:425]: dimm_id is defined as uint8_t so no need to check for negative value.

**Review by garnermic — CHANGES_REQUESTED** [review]
Changes requested based on Meta review.

**Comment by SaraLin-wiwynn** [comment]
Hi @garnermic 
Thanks for your comments.
Based on your suggestions to modify the code and force update this branch.
Please help review it again, thank you.

File: meta-facebook/yv35-cl/src/platform/plat_dimm.c (hunk 2 above)
Comment (SaraLin-wiwynn) [reply meta-facebook/yv35-cl/src/platform/plat_dimm.c]: We add new definition "DIMM_ID_UNKNOWN" to check if dimm_id can't map.

File: common/dev/i3c_dimm.c (hunk 1 above)
Comment (SaraLin-wiwynn) [reply common/dev/i3c_dimm.c]: Correct this typo to "According to".

**Pull request closed without merging**